core/src/main/java/com/linecorp/armeria/common/util/TimeWindowPercentileSampler.java

```diff
@@ -0,0 +1,87 @@
+ */
+package com.linecorp.armeria.common.util;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.common.metric.MoreMeters;
+
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.distribution.DistributionStatisticConfig;
+import io.micrometer.core.instrument.distribution.HistogramSnapshot;
+import io.micrometer.core.instrument.distribution.TimeWindowPercentileHistogram;
+
+/**
+ * Sample if the value is less than the percentile of the values in the last window.
+ */
+final class TimeWindowPercentileSampler implements Sampler<Long> {
+
+    private final float percentile;
+    private final long windowLengthMillis;
+    private final TimeWindowPercentileHistogram histogram;
+    @VisibleForTesting
+    static long SNAPSHOT_UPDATE_MILLIS = 1000L;
+    private long lastSnapshotMillis;
+    private HistogramSnapshot histogramSnapshot;
+
+    private final Clock clock;
+    private final AtomicReference<Boolean> isTakingSnapshot = new AtomicReference<>(false);
+
+    TimeWindowPercentileSampler(float percentile, long windowLengthMillis) {
+        this(percentile, windowLengthMillis, Clock.SYSTEM);
+    }
+
+    @VisibleForTesting
+    TimeWindowPercentileSampler(float percentile, long windowLengthMillis, Clock clock) {
+        this.percentile = percentile;
+        this.windowLengthMillis = windowLengthMillis;
+
+        final DistributionStatisticConfig distributionStatisticConfig =
+                DistributionStatisticConfig.builder()
+                                           .percentilesHistogram(false)
+                                           .percentiles(percentile)
+                                           .expiry(Duration.ofMillis(windowLengthMillis))
+                                           .build()
+                                           .merge(MoreMeters.distributionStatisticConfig());
+        this.histogram = new TimeWindowPercentileHistogram(clock, distributionStatisticConfig, true);
+        this.histogramSnapshot = histogram.takeSnapshot(0, 0, 0);
+        this.clock = clock;
+        this.lastSnapshotMillis = clock.wallTime();
+    }
+
+    static TimeWindowPercentileSampler create(float percentile, long windowLengthMillis) {
+        return new TimeWindowPercentileSampler(percentile, windowLengthMillis);
+    }
+
+    @Override
+    public boolean isSampled(Long t) {
+        histogram.recordLong(t);
+
+        if (lastSnapshotMillis + SNAPSHOT_UPDATE_MILLIS <= clock.wallTime()) {
+            if (isTakingSnapshot.compareAndSet(false, true)) {
+                histogramSnapshot = histogram.takeSnapshot(0, 0, 0);
+                lastSnapshotMillis = clock.wallTime();
+                isTakingSnapshot.set(false);
+            }
+        }
+
+        final Double percentileValue = histogramSnapshot.percentileValues()[0].value();
```

Comment:
Oops, Kotlin habits 😆

Reply:
Hmm turns out it doesn't work this way. Unit tests just started failing. I believe .value() just returns an object thus I need it this way.

Or this needs to change
```
        return t >= percentileValue.longValue();
```

I don't know why it would it fail anyway.

Reply:
I see, it just rounds down with `.longValue()` which works better in my case because I don't really care about the decimal points. This causes maximum value to be never sampled because it is just an approximation 🤔